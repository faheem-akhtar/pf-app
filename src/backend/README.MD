This folder is intented to be used only on backend side of the project (nextjs)
None of this files should be bundleled to the client builds