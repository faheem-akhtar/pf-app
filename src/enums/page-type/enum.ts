export enum PageTypeEnum {
  propertySerp = 'Property Serp',
}
