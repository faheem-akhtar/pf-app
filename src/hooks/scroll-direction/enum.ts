export enum HooksScrollDirectionEnum {
  UP = 'up',
  DOWN = 'down',
}
