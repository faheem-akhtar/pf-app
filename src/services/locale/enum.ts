export enum LocaleEnum {
  en = 'en',
  ar = 'ar',
}
