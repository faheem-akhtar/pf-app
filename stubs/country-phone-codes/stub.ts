import { CountryPhoneCodesInterface } from 'types/country-phone-codes/interface';

export const countryPhoneCodesStub = (data: Partial<CountryPhoneCodesInterface> = {}): CountryPhoneCodesInterface => ({
  code: 'ae',
  name: 'United Arab Emirates',
  phoneCode: '+971',
  ...data,
});
