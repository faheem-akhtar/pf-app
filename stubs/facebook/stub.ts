export const facebookStub = (statusResponse: { authResponse: { accessToken: string } }): typeof FB => {
  return {
    init: jest.fn(),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars
    login: ((callback, options) => callback(statusResponse as any)) as typeof FB.login,
  } as unknown as typeof FB;
};
