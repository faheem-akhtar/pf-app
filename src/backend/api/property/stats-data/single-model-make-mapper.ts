import { StatsDataPropertyInterface } from '@propertyfinder/pf-frontend-common/dist/module/stats/data/property.interface';
import { PropertyCompletionStatusType } from '@propertyfinder/pf-frontend-common/dist/module/stats/types';

import { BackendModelPropertyCategoryIdentifierEnum } from 'backend/model/property/category-identifier.enum';
import { BackendModelPropertyInterface } from 'backend/model/property/interface';
import { PROPERTY_SERP_ITEMS_PER_PAGE } from 'components/property/serp/items-per-page.constant';
import { configCommon } from 'config/common';

const getListingStatusDisplayed = (property: BackendModelPropertyInterface): string => {
  return property.areaSpecialistProperty ? 'Area Specialist Property' : property.listing_level_label;
};

const isRental = (category: BackendModelPropertyCategoryIdentifierEnum): boolean => {
  const rentalCategories = [
    BackendModelPropertyCategoryIdentifierEnum.residentialForRent,
    BackendModelPropertyCategoryIdentifierEnum.commercialForRent,
  ];

  return rentalCategories.includes(category);
};

const getCompletionStatus = (completionStatus: string | null): PropertyCompletionStatusType => {
  // Back end will return either 'off_plan', off plan', 'completed' or NULL
  const completionStatusMapping: {
    [key: string]: PropertyCompletionStatusType;
  } = {
    off_plan: 'off_plan',
    'off plan': 'off_plan',
    completed: 'completed',
  };

  return completionStatusMapping[completionStatus as string] || null;
};

export const backendApiPropertyStatsDataSingleModelMakeMapper =
  (pageNumber: number) =>
  (propertyModel: BackendModelPropertyInterface, index: number): StatsDataPropertyInterface => ({
    id: parseInt(propertyModel.id, 10),
    referenceNumber: propertyModel.reference,
    bedrooms: String(propertyModel.bedroom_value),
    bathrooms: String(propertyModel.bathroom_value),
    agent: {
      id: propertyModel.agent ? parseInt(propertyModel.agent.id, 10) : 0,
      name: propertyModel.agent ? propertyModel.agent.name : '',
      isVerified: propertyModel.agent ? propertyModel.agent.verification_status === 'verified' : false,
    },
    broker: {
      id: propertyModel.broker ? parseInt(propertyModel.broker.id, 10) : 0,
      name: propertyModel.broker ? propertyModel.broker.name : '',
    },
    title: propertyModel.name,
    url: propertyModel.links.self,
    image: propertyModel.links.image_property,
    licenseRera: propertyModel.rera,
    offeringType: propertyModel.category_identifier,
    housingType: propertyModel.type_identifier,
    housingTypeId: Number(propertyModel.property_type?.id),
    listingStatus: propertyModel.listing_level,
    listingStatusDisplayed: getListingStatusDisplayed(propertyModel),
    rentalPeriod: isRental(propertyModel.category_identifier) ? propertyModel.price_period_label : '',
    rentalPeriodId: isRental(propertyModel.category_identifier) ? propertyModel.price_period_identifier : '',
    size: propertyModel.size,
    agentUserId: parseInt((propertyModel.agent && propertyModel.agent.user_id) || '0'),
    verified: propertyModel.verified ? 'Verified' : 'Not verified',
    qs: propertyModel.quality_score,
    rsp: propertyModel.rsp,
    rss: propertyModel.rss,
    furnished: propertyModel.furnished,
    images: propertyModel.meta.images_count,
    locations: (propertyModel.location_tree || []).map((location) => parseInt(location.id, 10)),
    getPriceOnApp: !!propertyModel.price_on_application,
    price: propertyModel.meta.price,
    priceTxt: propertyModel.meta.price_text,
    currency: configCommon.currencyCode,
    subtitle: propertyModel.meta.subtitle,
    mortgageRent: false,
    moveSouq: false,
    position: PROPERTY_SERP_ITEMS_PER_PAGE * (pageNumber - 1) + index + 1,
    positionPage: index + 1,
    isSmartAd: propertyModel.smart_ad ? 1 : 0,
    isCtsProperty: propertyModel.cts ? 1 : 0,
    isAreaSpecialistProperty: propertyModel.areaSpecialistProperty ? 1 : 0,
    completionStatus: getCompletionStatus(propertyModel.completion_status),
    isExclusive: propertyModel.exclusive,
    has360View: !!propertyModel.view_360,
    newProjects: !!propertyModel.new_projects,
  });
