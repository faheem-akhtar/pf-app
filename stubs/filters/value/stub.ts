import { FiltersValueFieldPricePeriodType } from 'components/filters/value/field/price-period.type';
import { FiltersValueFieldSortType } from 'components/filters/value/field/sort.type';
import { FiltersValueInterface } from 'components/filters/value/interface';
import { FiltersCategoryIdEnum } from 'enums/filters/category-id.enum';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';

export const filtersValueStub = (): FiltersValueInterface => ({
  [FiltersParametersEnum.locationsIds]: [],
  [FiltersParametersEnum.categoryId]: FiltersCategoryIdEnum.residentialForRent,
  [FiltersParametersEnum.propertyTypeId]: '',
  [FiltersParametersEnum.minBedroom]: '',
  [FiltersParametersEnum.maxBedroom]: '',
  [FiltersParametersEnum.minBathroom]: '',
  [FiltersParametersEnum.maxBathroom]: '',
  [FiltersParametersEnum.minPrice]: null,
  [FiltersParametersEnum.maxPrice]: null,
  [FiltersParametersEnum.furnishing]: '',
  [FiltersParametersEnum.minArea]: null,
  [FiltersParametersEnum.maxArea]: null,
  [FiltersParametersEnum.pricePeriod]: 'y' as FiltersValueFieldPricePeriodType,
  [FiltersParametersEnum.keyword]: '',
  [FiltersParametersEnum.amenities]: [],
  [FiltersParametersEnum.completionStatus]: '',
  [FiltersParametersEnum.paymentMethod]: '',
  [FiltersParametersEnum.utilitiesPriceType]: '',
  [FiltersParametersEnum.virtualViewings]: '',
  [FiltersParametersEnum.isDeveloperProperty]: false,
  [FiltersParametersEnum.minInstallmentYears]: null,
  [FiltersParametersEnum.maxInstallmentYears]: null,
  [FiltersParametersEnum.sort]: 'mr' as FiltersValueFieldSortType,
  [FiltersParametersEnum.pageNumber]: 1,
});
