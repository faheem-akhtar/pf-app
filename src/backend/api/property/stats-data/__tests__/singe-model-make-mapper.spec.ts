import { propertyBackendStub } from 'stubs/property/backend.stub';

import { BackendModelPropertyCategoryIdentifierEnum } from 'backend/model/property/category-identifier.enum';

import { backendApiPropertyStatsDataSingleModelMakeMapper } from '../single-model-make-mapper';

describe('backendApiPropertyStatsDataSingleModelMakeMapper', () => {
  it('should map default data', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(mapper(propertyBackendStub(), 5)).toMatchSnapshot();
  });

  it('should map give getListingStatusDisplayed when areaSpecialistProperty is set', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          areaSpecialistProperty: true,
        },
        5
      ).listingStatusDisplayed
    ).toEqual('Area Specialist Property');
  });

  it('should have empty rentalPeriod, rentalPeriodId, when buy', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          category_identifier: BackendModelPropertyCategoryIdentifierEnum.residentialForSale,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        rentalPeriod: '',
        rentalPeriodId: '',
      })
    );
  });

  it('should have agentUserId = 0, when no agent available', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          agent: undefined,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        agentUserId: 0,
        agent: {
          id: 0,
          name: '',
          isVerified: false,
        },
      })
    );
  });

  it('should have broker.id = 0, when no broker available', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          broker: undefined,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        broker: {
          id: 0,
          name: '',
        },
      })
    );
  });

  it('should give Verified, for verified listing', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          verified: true,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        verified: 'Verified',
      })
    );
  });

  it('should handle empty location tree', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          location_tree: undefined,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        locations: [],
      })
    );
  });

  it('should map smart ad', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          smart_ad: true,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        isSmartAd: 1,
      })
    );
  });

  it('should map cts', () => {
    const mapper = backendApiPropertyStatsDataSingleModelMakeMapper(10);

    expect(
      mapper(
        {
          ...propertyBackendStub(),
          cts: true,
        },
        5
      )
    ).toEqual(
      expect.objectContaining({
        isCtsProperty: 1,
      })
    );
  });
});
