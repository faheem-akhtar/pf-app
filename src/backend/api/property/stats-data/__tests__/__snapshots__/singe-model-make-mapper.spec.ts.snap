// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backendApiPropertyStatsDataSingleModelMakeMapper should map default data 1`] = `
Object {
  "agent": Object {
    "id": 156740,
    "isVerified": false,
    "name": "Mr Hamid",
  },
  "agentUserId": 28140,
  "bathrooms": "1",
  "bedrooms": "0",
  "broker": Object {
    "id": 2867,
    "name": "Moon Star Real Estate",
  },
  "completionStatus": null,
  "currency": "AED",
  "furnished": "YES",
  "getPriceOnApp": false,
  "has360View": false,
  "housingType": "apartment",
  "housingTypeId": 1,
  "id": 7750053,
  "image": "https://www.propertyfinder.ae/property/3656619107f9c0c1f5eb8ae1539b9c81/500/356/MODE/2627b3/7750053-d109eo.jpg?ctr=ae",
  "images": 10,
  "isAreaSpecialistProperty": 0,
  "isCtsProperty": 0,
  "isExclusive": false,
  "isSmartAd": 0,
  "licenseRera": "",
  "listingStatus": "standard",
  "listingStatusDisplayed": "standard",
  "locations": Array [
    6,
    299,
    1933,
  ],
  "mortgageRent": false,
  "moveSouq": false,
  "newProjects": false,
  "offeringType": "rent",
  "position": 231,
  "positionPage": 6,
  "price": 3300,
  "priceTxt": "3,300 AED/month",
  "qs": 100,
  "referenceNumber": "FURNISH 33K",
  "rentalPeriod": "Monthly",
  "rentalPeriodId": "monthly",
  "rsp": 89.69,
  "rss": 61.19,
  "size": 650,
  "subtitle": "Apartment for Rent in Khalifa City A, Khalifa City",
  "title": "Family Community Furnish Studio/3300 M W/Markert",
  "url": "https://www.propertyfinder.ae/en/rent/apartment-for-rent-abu-dhabi-khalifa-city-khalifa-city-a-7750053.html",
  "verified": "Not verified",
}
`;
